contracts/EVMTokenBag.slb

```diff
@@ -20,79 +25,143 @@ library EVMTokenBag {
+        query.valueOrId = tokenId;
+        query.color = color;
+        (query, success) = findToken2(self, query, 10);
+        if (success) {
+            query.data = newData;
         }
     }
-    
-    function transfer(
+
+    function transferFrom(
         TokenBag memory self,
         address color,
         address from,
         address to,
-        uint value
+        uint valueOrId
     ) internal pure returns (bool) {
-        Output memory source;
-        Output memory dest;
-        // find source
+        if (to == address(0)) return false;
+        (Output memory source, bool success) = findToken(self, color, from, valueOrId);
+        // sender's token/tokens don't exist
+        if (!success) return false;
+        if (source.tokenType == ERC20TYPE) {
+            Output memory dest;
+            (dest, success) = findToken(self, color, to, valueOrId);
+            if (!success) {
+                (dest, success) = findToken(self, address(0), address(0), valueOrId);
+            }
+            // no slot found for output token
+            if (!success) return false;
+            // sender does not have enough tokens to send and/or underflow
+            if (source.valueOrId < valueOrId) return false;
+            // overflow
+            if (valueOrId + dest.valueOrId < valueOrId) return false;
+            
+            dest.owner = to;
+            dest.color = color;
+            dest.valueOrId += valueOrId;
+            source.valueOrId -= valueOrId;
+            
+            return true;
+        } else if (source.tokenType == ERC721TYPE || source.tokenType == ERC1948TYPE) {
+            source.owner = to;
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    function findToken(
+        TokenBag memory self,
+        address color,
+        address owner,
+        uint valueOrId
+    ) internal pure returns (Output memory, bool) {
+        Output memory token;
+        bool success;
+        bytes1 selector = 0x1f;
         for (uint i = 0; i < self.bag.length; i++) {
-            if (self.bag[i].owner == from && self.bag[i].color == color) {
-                source = self.bag[i];
-                break;
+            if (self.bag[i].owner == owner && self.bag[i].color == color) {
+                if (self.bag[i].tokenType == ERC20TYPE) {
+                    token = self.bag[i];
+                    success = true;
+                    break;
+                } else {
+                    if (self.bag[i].valueOrId == valueOrId) {
+                        token = self.bag[i];
+                        success = true;
+                        break;
+                    }
+                }
             }
         }
-    
-        // sender does not have enough tokens to send
-        if (source.valueOrId < value) return false;
-    
-        // find dest and/or empty
-        uint emptyId = 1337;
+        return (token, success);
+    }
+
+    function findToken2(
```

Comment:
document the difference to `findToken`, including `selector` param :)

Reply:
Done now.